litellm/llms/snowflake/completion.py

```diff
@@ -0,0 +1,379 @@
+
+        if response.status_code != 200:
+            raise SnowflakeError(
+                status_code=response.status_code, message=response.text
+            )
+
+        ## LOGGING
+        logging_obj.post_call(
+            input=data["content"],
+            api_key=headers.get("Authorization"),
+            original_response=response.text,
+            additional_args={"complete_input_dict": data},
+        )
+
+        response = self._process_response(
+            model_response=model_response,
+            response=response,
+            encoding=encoding,
+            prompt=data["content"],
+            model=model,
+        )
+        return response
+
+    async def async_streaming(
+        self,
+        model: str,
+        api_base: str,
+        logging_obj,
+        headers: dict,
+        data: Optional[dict],
+        client=None,
+    ):
+        if client is None:
+            client = AsyncHTTPHandler(timeout=httpx.Timeout(timeout=600.0, connect=5.0))
+
+        response = await client.post(api_base, headers=headers, data=json.dumps(data))
+
+        if response.status_code != 200:
+            raise SnowflakeError(
+                status_code=response.status_code, message=response.text
+            )
+
+        completion_stream = response.aiter_lines()
+
+        streamwrapper = CustomStreamWrapper(
+            completion_stream=completion_stream,
+            model=model,
+            custom_llm_provider="snowflake",
+            logging_obj=logging_obj,
+        )
+        return streamwrapper
+
+    def completion(
+        self,
+        model: str,
+        messages: list,
+        api_base: str,
+        acompletion: str,
+        custom_prompt_dict: dict,
+        model_response: ModelResponse,
+        print_verbose: Callable,
+        encoding,
+        api_key,
+        logging_obj,
+        optional_params=None,
+        litellm_params=None,
+        logger_fn=None,
+        headers={},
+        client=None,
+    ):
+        headers = validate_environment(api_key, headers)
+        if model in custom_prompt_dict:
+            # check if the model has a registered custom prompt
+            model_prompt_details = custom_prompt_dict[model]
+            prompt = custom_prompt(
+                role_dict=model_prompt_details["roles"],
+                initial_prompt_value=model_prompt_details["initial_prompt_value"],
+                final_prompt_value=model_prompt_details["final_prompt_value"],
+                messages=messages,
+            )
+        else:
+            prompt = prompt_factory(
+                model=model, messages=messages, custom_llm_provider="snowflake"
+            )
+
+        ## Load Config
+        config = SnowflakeTextConfig.get_config()
+        for k, v in config.items():
+            if k not in optional_params:
+                optional_params[k] = v
+
+        data = {
+            "model": model,
+            "messages": [{"content": prompt}],
```

Comment:
i'm confused - why does this convert the message list to a prompt and then back to a message list? 

if it's just to make sure content is a string, then i believe it might be better handled like Azure AI - https://github.com/BerriAI/litellm/blob/7630680690fdb85a0c239c49fa9817793b55447d/litellm/llms/azure_ai/chat/transformation.py#L27

Reply:
the message transformation logic should also be handled in a separate `transformation.py` file, so it's easy for someone to see how their call is being modified  - e.g. https://github.com/BerriAI/litellm/tree/main/litellm/llms/azure_ai